### ISSUE TITLE
SWIG tool does not correctly handle directors or quoted module names

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#2278]
This issue was originally created at: 2008-12-17 11:31:44.
This issue was reported by: `benmwebb`.
benmwebb said at 2008-12-17 11:31:44
>Consider the following simple C++ code and SWIG interface, and an SConstruct to
>build everything (may need to be adjusted on your system if Python lives
>somewhere else):
>test.h:
>class Foo {};
> 
>test.i:
>%module(directors="1") "test"
>%{
>#include "test.h"
>%}
>%include "test.h"
> 
>SConstruct:
>SharedObject("test", "test.i", SWIGFLAGS="-c++ -python",
>             CPPPATH="/usr/include/python2.5")
> 
>Note that in SWIG's %module directive we double-quote "test". This gives the
>same results from SWIG as calling the module test (i.e. without quotes).
>However, it confuses the SCons parser into thinking the generated Python file is
>called "test".py, which doesn't exist, so even after a successful run, "scons
>--debug=explain" reruns SWIG, reporting:
>$ scons --debug=explain
>scons: Reading SConscript files ...
>scons: done reading SConscript files.
>scons: Building targets ...
>scons: building `"test".py' because it doesn't exist
>swig -o test_wrap.cc -c++ -python test.i
>scons: done building targets.
> 
> 
>The second problem is that when SWIG is given the 'directors' module argument,
>it additionally generates a header, in this case test_wrap.h, which is included
>by test_wrap.cc. SCons doesn't know about this header, so won't clean it up at
>'scons -c' time, and does additional unnecessary builds because of it:
>$ scons --debug=explain
>scons: Reading SConscript files ...
>scons: done reading SConscript files.
>scons: Building targets ...
>scons: rebuilding `test.os' because `test_wrap.h' is a new dependency
>g++ -o test.os -c -fPIC -I/usr/include/python2.5 test_wrap.cc
>scons: done building targets.
> 
>I will attach a patch that fixes both of these issues for me. Please let me know
>if you have any questions.

benmwebb said at 2008-12-17 11:32:47
>Created an attachment (id=548)
>scons-swig.patch

benmwebb said at 2008-12-17 11:34:41
>BTW, I ran into this with the Fedora 10 package of SCons, which claims to be
>scons-1.0.0-1.d20080826.fc10.noarch, but I just checked with the 1.1.0.d20081207
>checkpoint release, and Tool/swig.py is the same in that release, so I have
>every reason to suspect the bugs are still present and that this patch will fix
>them.

benmwebb said at 2008-12-17 12:03:29
>Created an attachment (id=549)
>scons-swig2.patch

benmwebb said at 2008-12-17 12:05:43
>Actually I see that the first patch doesn't work correctly in all cases. Please
>check out scons-swig2.patch instead. The issue is that SWIG generates a header
>file only once for each test_wrap.cc file, *not* for each module, so if for some
>reason your module name does not match the name of the .i file, my original
>patch doesn't work. The revised patch adds a header based on the name of the
>original _wrap.cc file. It seems a little inelegant though.

gregnoel said at 2008-12-26 14:58:04
>Assigning this issue to you {;-} with me as QA.  Your patch is a good one, but
>we like to only have one fix per patch as it makes it easier to identify what
>went wrong in the event of a problem.  It's probably not absolutely necessary in
>this case, but it's good practice, so if you could break the patch up into two
>pieces and review them on the dev list, I'll apply them for you.  Tks.

gregnoel said at 2008-12-26 14:59:38
>Oops, forgot the milestone.  We'll try to squeeze this in to 1.3 if we can.

benmwebb said at 2009-01-03 14:19:21
>Created an attachment (id=561)
>swig-module-quoted.patch; a fix for the "quoted module name" part of this issue

gregnoel said at 2009-01-04 00:27:28
>The "quoted module name" patch has been applied as revision 3872.

benmwebb said at 2009-01-05 22:06:12
>Created an attachment (id=562)
>swig-director-header.patch; fix for generation of foo_wrap.h header

benmwebb said at 2009-01-10 17:24:02
>Hmm, this last patch breaks SCons for old Python versions. To fix this, replace
>the following line in the patched file:
> 
>        directors = directors or 'directors' in m[0]
> 
>with:
> 
>        directors = directors or m[0].find('directors') >= 0

benmwebb said at 2009-01-13 17:54:24
>Created an attachment (id=569)
>swig-director-header2.patch; fixed patch with Python 1.5.2 support and better tests

gregnoel said at 2009-01-17 12:05:46
>Slight modification of "director header" patch submitted as revision 3904.  This
>concludes the action on this issue.

gregnoel said at 2010-07-16 07:12:34
>*** Issue 2171 has been marked as a duplicate of this issue. ***

benmwebb attached [scons-swig.patch](https://github.com/SCons/tigris-issue-attachments/blob/master/2278/548/scons-swig.patch) at 2008-12-17 11:32:47.
>scons-swig.patch

benmwebb attached [scons-swig2.patch](https://github.com/SCons/tigris-issue-attachments/blob/master/2278/549/scons-swig2.patch) at 2008-12-17 12:03:29.
>scons-swig2.patch

benmwebb attached [swig-module-quoted.patch](https://github.com/SCons/tigris-issue-attachments/blob/master/2278/561/swig-module-quoted.patch) at 2009-01-03 14:19:21.
>swig-module-quoted.patch; a fix for the "quoted module name" part of this issue

benmwebb attached [swig-director-header.patch](https://github.com/SCons/tigris-issue-attachments/blob/master/2278/562/swig-director-header.patch) at 2009-01-05 22:06:12.
>swig-director-header.patch; fix for generation of foo_wrap.h header

benmwebb attached [swig-director-header2.patch](https://github.com/SCons/tigris-issue-attachments/blob/master/2278/569/swig-director-header2.patch) at 2009-01-13 17:54:24.
>swig-director-header2.patch; fixed patch with Python 1.5.2 support and better tests

gregnoel said this issue is duplicated by #2171 at 2010-07-16 07:12:34.
